include/swift/Sema/ConstraintSystem.h

```diff
@@ -4623,7 +4623,10 @@ class ConstraintSystem {
   /// bindings, e.g., the supertypes of a given type.
   struct PotentialBinding {
     /// The type to which the type variable can be bound.
-    Type BindingType;
+    ///
+    /// Note that the type is mutable since it could be
+    /// adjusted as a result of type-join operation.
+    mutable Type BindingType;
```

Comment:
@hborla I think we can actually redo this now. Currently we-update the type but don't update the type cache. I think instead we should remove and re-introduce the binding at the same spot.

Reply:
Done!